python-lib/generalized_linear_models/dku_glm.py

```diff
@@ -0,0 +1,230 @@
+        y_pred = np.array(self.fitted_model.predict(X))
+        y_pred_final = y_pred.reshape((len(y_pred), -1))
+        
+        return y_pred_final>0.5
+           
+
+    def predict_proba(self, X):
+        """
+        Return the prediction proba
+        """
+        #  adds a constant 
+        X = sm.add_constant(X, has_constant='add')
+        
+        # makes predictions and converts to DSS accepted format 
+        y_pred = np.array(self.fitted_model.predict(X))
+        y_pred_final = y_pred.reshape((len(y_pred), -1))
+        
+        # returns p, 1-p prediction probabilities
+        return np.append(1-y_pred_final, y_pred_final, axis=1)
+    
+    
+    
+    
+    
+class RegressionGLM(BaseEstimator, ClassifierMixin):
+    
+    def __init__(self, link, family, alpha, power, var_power):
+
+        self.link = link
+        self.family = family
+        self.alpha = alpha
+        self.power = power
+        self.var_power = var_power
+        self.fit_intercept=True
+        self.intercept_scaling = 1
+        self.fitted_model = None
+        self.coef_= None
+        self.intercept_ = None
+        self.classes_ = None
+
+        
+    def get_link(self):
+        """ 
+        gets the statsmodel link function based on the
+        user defined link on the model training screen   
+        """ 
+        links_dict = {
+            'cloglog': sm.families.links.cloglog(),
+            'log': sm.families.links.log(),
+            'logit': sm.families.links.logit(),
+            'negative_binomial': sm.families.links.NegativeBinomial(self.alpha),
+            'power': sm.families.links.Power(self.power),
+            'cauchy': sm.families.links.cauchy(),
+            'identity': sm.families.links.identity(),
+            'inverse_power': sm.families.links.inverse_power(),
+            'inverse_squared': sm.families.links.inverse_squared()
+        }
+        
+        return links_dict[self.link]
+        
+    def get_family(self, link):
+        """ 
+        takes in user defined family variable 
+        and statsmodel link function
+        returns the family 
+        """       
+        if self.family == 'binomial':
+            return sm.families.Binomial(link=link)
+        
+        elif self.family == "gamma":
+            return sm.families.Gamma(link=link)
+        
+        elif self.family == "gaussian":
+            return sm.families.Gaussian(link=link)
+        
+        elif self.family == "inverse_gaussian": 
+            return sm.families.InverseGaussian(link=link) 
+        
+        elif self.family == "negative_binomial": 
+            return sm.families.NegativeBinomial(link=link, alpha=self.alpha)
+        
+        elif self.family == "poisson":
+            return sm.families.Poisson(link=link)
+        
+        elif self.family == "tweedie":
+            return sm.families.Tweedie(link=link, var_power=self.var_power)
+        else:
+            exit()
```

Comment:
I had missed this one:
```suggestion
            raise ValueError("Unsupported family")
```
The real solution would be to use class inheritance. Thanks to this, you will be free from code duplication :D

Reply:
d56fdc4b311f3e3e121f6a3c9d47eb61ebefc64e